syntax = "proto2";

message Data {
	required string id = 1;
	required int32 num_train = 2;
	required int32 num_test = 3;
	repeated string label_name = 4;
	optional Source source = 5;

	message Source{
		required string train_path = 1;
		required string test_path = 2;
		required string val_path = 4;
		required Type format = 3;
		enum Type{
			H5=0;
			LMDB=1;
			NUMPY=2;
		}
	}
}

message Logit{
	repeated float data = 1 [packed=true];
}

message ClassifierRawData
{
	repeated Logit logits = 1;
    repeated int32 gt = 2 [packed=true];
	repeated int32 id = 3 [packed=true];
}

message PerformanceMetrics
{
    optional double time = 1 [default=0];
    optional int64 params = 2;
    optional int64 ops = 3;
}

message Classifier {
	required string id = 1;
    optional string classifier_file = 2;
    optional string component_id = 3;
    optional string data_id = 4;
}

message Merger {

	required string id = 1;
	enum MergeType{
		AVERAGE = 0;
		MAX = 1;
		VOTING = 2;
		ADABOOST_LABEL_WEIGHTS_LOGIT_INFERENCE = 3;
		ADABOOST_LABEL_WEIGHTS_LABEL_INFERENCE = 4;
		ADABOOST_LOGIT_WEIGHTS_LABEL_INFERENCE = 5;
		ADABOOST_LOGIT_WEIGHTS_LOGIT_INFERENCE = 6;
		ADABOOST_LOGIT_WEIGHTS_MAX_INFERENCE = 7;
		ADABOOST_LABEL_WEIGHTS_MAX_INFERENCE = 8;
	}
	required MergeType merge_type = 2 [default = AVERAGE];
	repeated string component_ids = 3;
}

message Trigger {
	required string id = 1;
	optional Classifier classifier = 2;
	optional string model = 3;
	repeated string component_ids = 4;
}


message System {
	repeated Data data = 1;
	repeated Trigger trigger = 2;
	repeated Classifier classifier = 3;
    repeated Merger merger = 4;
}
